### ISSUE TITLE
Support TripletLoss

### ISSUE THREAD

**Issue by chengtbf** [Oneflow-Inc/oneflow#3630]
TripletLoss is a common loss function. OneFlow is not supported at present. 

It is necessary to add forward and **backward** implementations of this loss.

reference implementation using PyTorch:  https://github.com/Cysu/open-reid/blob/master/reid/loss/triplet.py
```
class TripletLoss(nn.Module):
    def __init__(self, margin=0):
        super(TripletLoss, self).__init__()
        self.margin = margin
        self.ranking_loss = nn.MarginRankingLoss(margin=margin)

    def forward(self, inputs, targets):
        n = inputs.size(0)
        # Compute pairwise distance, replace by the official when merged
        dist = torch.pow(inputs, 2).sum(dim=1, keepdim=True).expand(n, n)
        dist = dist + dist.t()
        dist.addmm_(1, -2, inputs, inputs.t())
        dist = dist.clamp(min=1e-12).sqrt()  # for numerical stability
        # For each anchor, find the hardest positive and negative
        mask = targets.expand(n, n).eq(targets.expand(n, n).t())
        dist_ap, dist_an = [], []
        for i in range(n):
            dist_ap.append(dist[i][mask[i]].max())
            dist_an.append(dist[i][mask[i] == 0].min())
        dist_ap = torch.cat(dist_ap)
        dist_an = torch.cat(dist_an)
        # Compute ranking hinge loss
        y = dist_an.data.new()
        y.resize_as_(dist_an.data)
        y.fill_(1)
        y = Variable(y)
        loss = self.ranking_loss(dist_an, dist_ap, y)
        prec = (dist_an.data > dist_ap.data).sum() * 1. / y.size(0)
        return loss, prec
```

**Comment by MARD1NO** [comment]
郑泽康接手

**Comment by chengtbf** [comment]
> 郑泽康接手

这个任务的进展如何？ @MARD1NO

**Comment by kanagi2587** [comment]
请问这个有新进展了么，我手中的算法目前也需要这个损失函数
